atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java

```diff
@@ -514,6 +520,175 @@ public void setOnceTest() {
+
+        keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.multipleCells(
+                TEST_TABLE,
+                firstTestCell.getRowName(),
+                ImmutableMap.of(firstTestCell, firstVal),
+                ImmutableMap.of(nextTestCell, secondVal)));
+        verifyMultiCheckAndSet(ImmutableMap.of(firstTestCell, firstVal, nextTestCell, secondVal));
+    }
+
+    @Test
+    public void testMcasChecksAllColumnsIfExpectedIsEmpty() {
+        Cell firstTestCell = Cell.create(row(0), column(0));
+        Cell nextTestCell = Cell.create(row(0), column(1));
+
+        byte[] firstVal = val(0, 0);
+        byte[] secondVal = val(0, 1);
+        keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                TEST_TABLE,
+                firstTestCell.getRowName(),
+                ImmutableMap.of(firstTestCell, firstVal, nextTestCell, firstVal)));
+        verifyMultiCheckAndSet(ImmutableMap.of(firstTestCell, firstVal, nextTestCell, firstVal));
+
+        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                        TEST_TABLE, firstTestCell.getRowName(), ImmutableMap.of(nextTestCell, secondVal))))
+                .isInstanceOf(MultiCheckAndSetException.class);
+    }
+
+    @Test
+    public void testMultiCheckAndSetCannotUpdateAcrossMultipleRows() {
+        Cell firstTestCell = Cell.create(row(0), column(0));
+        Cell nextTestCell = Cell.create(row(1), column(1));
+
+        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                        TEST_TABLE,
+                        firstTestCell.getRowName(),
+                        ImmutableMap.of(firstTestCell, val(0, 0), nextTestCell, val(0, 1)))))
+                .isInstanceOf(SafeIllegalStateException.class)
+                .hasMessageContaining("Can only update cells in one row.");
+    }
+
+    @Test
+    public void testMultiCheckAndSetIndependentlyFails() {
+        Cell firstTestCell = Cell.create(row(0), column(0));
+        Cell nextTestCell = Cell.create(row(1), column(1));
+
+        keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                TEST_TABLE, firstTestCell.getRowName(), ImmutableMap.of(firstTestCell, val(0, 0))));
+        keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                TEST_TABLE, nextTestCell.getRowName(), ImmutableMap.of(nextTestCell, val(0, 1))));
+
+        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                        TEST_TABLE, nextTestCell.getRowName(), ImmutableMap.of(nextTestCell, val(0, 2)))))
+                .isInstanceOf(MultiCheckAndSetException.class);
+
+        verifyMultiCheckAndSet(ImmutableMap.of(firstTestCell, val(0, 0), nextTestCell, val(0, 1)));
+    }
+
+    private void verifyMultiCheckAndSet(Map<Cell, byte[]> expectedValues) {
```

Comment:
We might as well test a multicas that puts a new cell and updates an existing one -- we won't use this in txn4, but we do want to know that it works

Reply:
sure added an explicit test - I think the thing to _really_ know is that the "check" will only happen on the columns explicitly declared as expected and on all columns, if no columns are declared. I will add this to the doc.